pkg/scale/decode.go

```diff
@@ -0,0 +1,539 @@
+	err = ds.unmarshal(dstv.Elem())
+	if err != nil {
+		return
+	}
+	return
+}
+
+type decodeState struct {
+	bytes.Buffer
+}
+
+func (ds *decodeState) unmarshal(dstv reflect.Value) (err error) {
+	in := dstv.Interface()
+	switch in.(type) {
+	case *big.Int:
+		err = ds.decodeBigInt(dstv)
+	case *Uint128:
+		err = ds.decodeUint128(dstv)
+	case int, uint:
+		err = ds.decodeUint(dstv)
+	case int8, uint8, int16, uint16, int32, uint32, int64, uint64:
+		err = ds.decodeFixedWidthInt(dstv)
+	case []byte:
+		err = ds.decodeBytes(dstv)
+	case string:
+		err = ds.decodeBytes(dstv)
+	case bool:
+		err = ds.decodeBool(dstv)
+	default:
+		switch reflect.TypeOf(in).Kind() {
+		case reflect.Ptr:
+			err = ds.decodePointer(dstv)
+		case reflect.Struct:
+			t := reflect.TypeOf(in)
+			// check if this is a convertible to Result, if so encode using decodeResult
+			switch t.ConvertibleTo(reflect.TypeOf(Result{})) {
+			case true:
+				err = ds.decodeResult(dstv)
+			case false:
+				err = ds.decodeStruct(dstv)
+			}
+		case reflect.Array:
+			err = ds.decodeArray(dstv)
+		case reflect.Slice:
+			t := reflect.TypeOf(in)
+			// check if this is a convertible to VaryingDataType, if so encode using encodeVaryingDataType
+			switch t.ConvertibleTo(reflect.TypeOf(VaryingDataType{})) {
+			case true:
+				err = ds.decodeVaryingDataType(dstv)
+			case false:
+				err = ds.decodeSlice(dstv)
+			}
+		default:
+			_, ok := in.(VaryingDataTypeValue)
+			switch ok {
+			case true:
+				var temp reflect.Value
+				t := reflect.TypeOf(in)
+				switch t.Kind() {
+				// TODO: support more primitive types.  Do we need to support arrays and slices as well?
+				case reflect.Int:
+					temp = reflect.New(reflect.TypeOf(int(1)))
+					err = ds.unmarshal(temp.Elem())
+					if err != nil {
+						break
+					}
+				case reflect.Int16:
+					temp = reflect.New(reflect.TypeOf(int16(1)))
+					err = ds.unmarshal(temp.Elem())
+					if err != nil {
+						break
+					}
+				default:
+					err = fmt.Errorf("unsupported kind for VaryingDataTypeValue: %s", t.Kind())
+					return
+				}
+				dstv.Set(temp.Elem().Convert(t))
+			default:
+				err = fmt.Errorf("unsupported type: %T", in)
+			}
+		}
+	}
+	return
+}
+
+func (ds *decodeState) decodeResult(dstv reflect.Value) (err error) {
+	dstt := reflect.TypeOf(dstv.Interface())
+	key := fmt.Sprintf("%s.%s", dstt.PkgPath(), dstt.Name())
+	resultCases, ok := resCache[key]
+	if !ok {
+		err = fmt.Errorf("unable to find registered custom Result: %T", dstv.Interface())
+		return
+	}
+
+	var rb byte
+	rb, err = ds.ReadByte()
+	if err != nil {
+		return
+	}
+	switch rb {
+	case 0x00:
+		// Ok case
+		okIn, ok := resultCases[true]
+		if !ok {
+			err = fmt.Errorf("unable to find registered custom Result.Ok for: %T", dstv.Interface())
+			return
+		}
+		newOk := reflect.New(reflect.TypeOf(okIn))
+		err = ds.unmarshal(newOk.Elem())
+		if err != nil {
+			break
+		}
+		res := reflect.New(dstt)
+		res.Elem().FieldByName("Ok").Set(newOk.Elem())
```

Comment:
should this do a check for the "Ok" field? or is it checked when the type is registered?

Reply:
`decodeResult` is called with a `reflect.Value` of type `Result`.  Technically this would panic if a `Result` isn't passed in, but that shouldn't happen, and this is a private receiver function.